### DISCUSSION TITLE
Sonoff Basic  - Time Delay after switch off?

### DISCUSSION THREAD

**Discussion by homestead72** [arendst/Tasmota#14217] [Q&A] [answered]
Hello all,

I am trying to get a Sonoff Basic (flashed with V10 Tasmota) to stay on for a certain time delay after switching off.  The code in the Tasmota docs suggests the following rule:

_Time-delay After Switch Off~
Rule~
Backlog switchmode1 1; rule1 1

Rule1 ON switch1#state=1 DO Backlog Power1 on; ruletimer1 0 ENDON
ON switch1#state=0 DO ruletimer1 300 ENDON
ON rules#timer=1 DO Power1 0 ENDON_

But I cannot get this to work at all.  When I switch the Sonoff off it just goes off immediatley.  My head hurts!  What am I doing wrong please?

Thanks in advance

Stuart

**Comment by barbudor** [comment, marked as answer]
The rule looks ok (given the _ at the end is not part of the rule which in that case would break the bahavior)

How did you connected the switch ? On the internal header ? which GPIO ?
Is that a rocker switch with 2 stable position (ON contact closed, OFF contact open)  and not a push-button ?

**Reply by homestead72** [reply]
Thanks for your reply @barbudor 

It would seem I have completely missed the point somewhere.  I wanted to the time delay to occur if I power off the sonoff basic either by the web interface or by the manual button on the device.  If I need to have an external does this need to go onto GPIO14 and if so where do I find that on a Sonoff Basic r2?

**Reply by barbudor** [reply]
You stated that you wanted the rule to react on a switch change. A "Switch" is an external component so yes it needs to be configured on a GPIO.

Now, as you give more details:
1) Adding a delay on web button action is not possible as it is equivalent to the POWER command
2) Adding a delay on the existing hardware push button is possible but more complicated than with a switch
3) Adding a delay on a MQTT Power message is not possible as it is equivalent as the POWER  command


=> 2) The existing push button only provide a single event TOGGLE (=2) which is independant of the current power state, while a rocker switch provide 2 distinct ON (=1) or OFF (=0) states.
So you need a more complicated rule which knows the current power state.
If you could hook a rocker switch to GPIO14, that would be simpler

=> 3) You can create an EVENT that can be called by an external MQTT system such as MyPower 
```
ON event#MyPower=1 DO backlog Power 1; RuleTimer1 0 ENDON
ON event#MyPower=0 DO RuleTime1 300 ENDON
ON Rules#Timer=1 DO Power 0 ENDON
```
Can be triggered through MQTT by publishing to `cmnd/yourtopic/event` with payload `MyPower=0` or `MyPower=1`

**Reply by barbudor** [reply]
![image](https://user-images.githubusercontent.com/2996042/147755467-d596f38e-55f1-4ce9-abaf-631be066f894.png)

**Reply by homestead72** [reply]
Many thanks @barbudor.  I have successfully fitted an external switch and got the time delay working.

**Comment by sfromis** [comment]
One setup to handle actions from commands, web UI and physical button at the same time could be to configure a "virtual" relay on an unused pin, and have this be the subject of all the actions. And then have rules to react to the state changes of that, and control the real relay as appropriate, including whatever timing setup you want. This way, you need no "special" commands like Event, but can have things work as normal, just with a bit of extra control in the rules.
